# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) Pandas for Exploratory Data Analysis I 
by [@josephofiowa](https://twitter.com/josephofiowa)

Pandas is the most prominent Python library for exploratory data analysis (EDA). The functions Pandas supports are integral to understanding, formatting, and preparing our data. Formally, we use Pandas to investigate, wrangle, munge, and clean our data. Pandas is the Swiss Army Knife of data manipulation!


We'll have two coding-heavy sessions on Pandas. In this one, we'll use Pandas to:
 - Read in a dataset
 - Investigate a dataset's integrity
 - Filter, sort, and manipulate a DataFrame's series

## About the Dataset: Adventureworks Cycles

<img align="right" src="http://lh6.ggpht.com/_XjcDyZkJqHg/TPaaRcaysbI/AAAAAAAAAFo/b1U3q-qbTjY/AdventureWorks%20Logo%5B5%5D.png?imgmax=800">

For today's Pandas exercises, we will be using a dataset developed by Microsoft for training purposes in SQL server, known the [Adventureworks Cycles 2014OLTP Database](https://github.com/Microsoft/sql-server-samples/releases/tag/adventureworks). It is based on a fictitious company called Adventure Works Cycles (AWC), a multinational manufacturer and seller of bicycles and accessories. The company is based in Bothell, Washington, USA and has regional sales offices in several countries. We will be looking at a single table from this database, the Production.Product table, which outlines some of the products this company sells. 

A full data dictionary can be viewed [here](https://www.sqldatadictionary.com/AdventureWorks2014/).


Let's take a closer look at the Production.Product table [data dictionary](https://www.sqldatadictionary.com/AdventureWorks2014/Production.Product.html), which is a description of the fields (columns) in the table (the .csv file we will import below):
- **ProductID** - Primary key for Product records.
- **Name** - Name of the product.
- **ProductNumber** - Unique product identification number.
- **MakeFlag** - 0 = Product is purchased, 1 = Product is manufactured in-house.
- **FinishedGoodsFlag** - 0 = Product is not a salable item. 1 = Product is salable.
- **Color** - Product color.
- **SafetyStockLevel** - Minimum inventory quantity.
- **ReorderPoint** - Inventory level that triggers a purchase order or work order.
- **StandardCost** - Standard cost of the product.
- **ListPrice** - Selling price.
- **Size** - Product size.
- **SizeUnitMeasureCode** - Unit of measure for the Size column.
- **WeightUnitMeasureCode** - Unit of measure for the Weight column.
- **DaysToManufacture** - Number of days required to manufacture the product.
- **ProductLine** - R = Road, M = Mountain, T = Touring, S = Standard
- **Class** - H = High, M = Medium, L = Low
- **Style** - W = Womens, M = Mens, U = Universal
- **ProductSubcategoryID** - Product is a member of this product subcategory. Foreign key to ProductSubCategory.ProductSubCategoryID.
- **ProductModelID** - Product is a member of this product model. Foreign key to ProductModel.ProductModelID.
- **SellStartDate** - Date the product was available for sale.
- **SellEndDate** - Date the product was no longer available for sale.
- **DiscontinuedDate** - Date the product was discontinued.
- **rowguid** - ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample.
- **ModifiedDate** - Date and time the record was last updated.


## Importing Pandas

To [import a library](https://docs.python.org/3/reference/import.html), we write `import` and the library name. For Pandas, is it common to name the library `pd` so that when we reference a function from the Pandas library, we only write `pd` to reference the aliased [namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) -- not `pandas`.

In [1]:
import pandas as pd

In [2]:
# we can see the details about the imported package by referencing its private class propertys:
print(f'I am using {pd.__name__} \
Version: {pd.__version__}.\n\
It is installed at: {pd.__path__}')

I am using pandas Version: 0.24.2.
It is installed at: ['//anaconda3/lib/python3.7/site-packages/pandas']


## Reading in Data

Pandas dramatically simplifies the process of reading in data. When we say "reading in data," we mean loading a file into our machine's memory.

When you have a CSV, for example, and then you double-click to open it in Microsoft Excel, the open file is "read into memory." You can now manipulate the CSV.

When we read data into memory in Python, we are creating an object. We will soon explore this object. _(And, as an aside, when we have a file that is greater than the size of our computer's memory, this is approaching "Big Data.")_

Because we are working with a CSV, we will use the [read CSV](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method.<br>A [delimiter](https://en.wikipedia.org/wiki/Delimiter-separated_values) is a character that separates fields (columns) in the imported file. Just because a file says `.csv` does not necessarily mean that a comma is used as the delimiter. In this case, we have a tab character as the delimiter for our columns, so we will be using `sep='\t'` to tell pandas to 'cut' the columns every time it sees a [tab character in the file](http://vim.wikia.com/wiki/Showing_the_ASCII_value_of_the_current_character).

In [3]:
# read the dataset as a DataFrame into a variable named 'prod'
prod = pd.read_csv("../data/Production.Product.csv", sep = "\t")

In [4]:
# show the head of this dataset
prod.head()

ProductID                   Name ProductNumber  MakeFlag  \
0          1        Adjustable Race       AR-5381         0   
1          2           Bearing Ball       BA-8327         0   
2          3        BB Ball Bearing       BE-2349         1   
3          4  Headset Ball Bearings       BE-2908         0   
4        316                  Blade       BL-2036         1   

   FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  StandardCost  \
0                  0   NaN              1000           750           0.0   
1                  0   NaN              1000           750           0.0   
2                  0   NaN               800           600           0.0   
3                  0   NaN               800           600           0.0   
4                  0   NaN               800           600           0.0   

   ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
0        0.0  ...         NaN   NaN   NaN                   NaN   
1        0.0  ...         NaN   NaN   NaN                   NaN   
2        0.0  ...         NaN   NaN   NaN                   NaN   
3        0.0  ...         NaN   NaN   NaN                   NaN   
4        0.0  ...         NaN   NaN   NaN                   NaN   

   ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
0             NaN  2008-04-30 00:00:00         NaN              NaN   
1             NaN  2008-04-30 00:00:00         NaN              NaN   
2             NaN  2008-04-30 00:00:00         NaN              NaN   
3             NaN  2008-04-30 00:00:00         NaN              NaN   
4             NaN  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  
3  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}  2014-02-08 10:01:36.827000000  
4  {E73E9750-603B-4131-89F5-3DD15ED5FF80}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

*Documentation Pause*

How did we know how to use `pd.read_csv`? Let's take a look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). Note the first argument required (`filepath`).
> Take a moment to dissect other arguments and options when reading in data.

We have just created a data structure called a `DataFrame`. See?

In [5]:
# show the type of this 'prod' dataset
type(prod)

pandas.core.frame.DataFrame

## Inspecting our DataFrame: The basics

We'll now perform basic operations on the DataFrame, denoted with comments.

In [6]:
# print the first and last 3 rows
prod.tail()

ProductID                Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
499        995   ML Bottom Bracket       BB-8107         1                  1   
500        996   HL Bottom Bracket       BB-9108         1                  1   
501        997  Road-750 Black, 44    BK-R19B-44         1                  1   
502        998  Road-750 Black, 48    BK-R19B-48         1                  1   
503        999  Road-750 Black, 52    BK-R19B-52         1                  1   

     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
499    NaN               500           375       44.9506     101.24  ...   
500    NaN               500           375       53.9416     121.49  ...   
501  Black               100            75      343.6496     539.99  ...   
502  Black               100            75      343.6496     539.99  ...   
503  Black               100            75      343.6496     539.99  ...   

    ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
499         NaN    M    NaN                   5.0            96.0   
500         NaN    H    NaN                   5.0            97.0   
501          R     L     U                    2.0            31.0   
502          R     L     U                    2.0            31.0   
503          R     L     U                    2.0            31.0   

           SellStartDate SellEndDate DiscontinuedDate  \
499  2013-05-30 00:00:00         NaN              NaN   
500  2013-05-30 00:00:00         NaN              NaN   
501  2013-05-30 00:00:00         NaN              NaN   
502  2013-05-30 00:00:00         NaN              NaN   
503  2013-05-30 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
499  {71AB847F-D091-42D6-B735-7B0C2D82FC84}  2014-02-08 10:01:36.827000000  
500  {230C47C5-08B2-4CE3-B706-69C0BDD62965}  2014-02-08 10:01:36.827000000  
501  {44CE4802-409F-43AB-9B27-CA53421805BE}  2014-02-08 10:01:36.827000000  
502  {3DE9A212-1D49-40B6-B10A-F564D981DBDE}  2014-02-08 10:01:36.827000000  
503  {AE638923-2B67-4679-B90E-ABBAB17DCA31}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

Notice that `.head()` is a method (denoted by parantheses), so it takes arguments.

**Class Question:** 
- What do you think changes if we pass a different number `head()` argument?
- How would we print the last 5 rows?

In [7]:
# identify the shape (rows by columns)
prod.shape

(504, 25)

Here we have 504 rows, and 25 columns. This is a tuple, so we can extract the parts using indexing:

In [8]:
# print the number of rows
prod.shape[0]

504

## Using the index
An [index](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.html) is an immutable ndarray implementing an ordered, sliceable set. The basic object storing axis labels for all pandas objects. Think of it as a 'row address' for your data frame (table). It is best practice to explicitly set the index of your dataframe, as these are commonly used as [primary keys](https://en.wikipedia.org/wiki/Primary_key) which can be used to [join](https://www.w3schools.com/sql/sql_join.asp) your dataframe to other dataframes.

The dataframe can store different types (int, string, datetime), and when importing data will automatically assign a number to each row, starting at zero and counting up. You can overwrite this, which is what we are going to do.

In [9]:
# display the index as it sits (auto-generated upon import)
prod.index

RangeIndex(start=0, stop=504, step=1)

In [10]:
# note that our auto-generated index has no name

In [11]:
# Here we are looking at three columns;
# the one on the left is the index (automatically generated upon import by pandas)
# 'ProductID' is our PK (primary key) from our imported table. 'Name' is a data column.
# Notice that the generated index starts at zero and our PK starts at 1.

In [10]:
# Setting the index overwrites the automatically generated index
# with our PK, which resided in the 'ProductID' column.
prod.set_index("ProductID", inplace = True)

In [11]:
# Note how our index property has changed as a result
prod.head()

Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   

          Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  \
ProductID                                                                       
1           NaN              1000           750           0.0        0.0  NaN   
2           NaN              1000           750           0.0        0.0  NaN   
3           NaN               800           600           0.0        0.0  NaN   
4           NaN               800           600           0.0        0.0  NaN   
316         NaN               800           600           0.0        0.0  NaN   

           ... ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID  ...                                                                 
1          ...         NaN   NaN    NaN                   NaN            NaN   
2          ...         NaN   NaN    NaN                   NaN            NaN   
3          ...         NaN   NaN    NaN                   NaN            NaN   
4          ...         NaN   NaN    NaN                   NaN            NaN   
316        ...         NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

In [12]:
# And our index has also inherited the name of our 'ProductID' column
prod.index

Int64Index([  1,   2,   3,   4, 316, 317, 318, 319, 320, 321,
            ...
            990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
           dtype='int64', name='ProductID', length=504)

## Column headers and datatypes

In [13]:
# print the columns
prod.columns

Index(['Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag', 'Color',
       'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size',
       'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight',
       'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object')

In [14]:
# examine the datatypes of the columns
# note that these were automatically inferred by pandas upon import!
prod.dtypes

Name                      object
ProductNumber             object
MakeFlag                   int64
FinishedGoodsFlag          int64
Color                     object
SafetyStockLevel           int64
ReorderPoint               int64
StandardCost             float64
ListPrice                float64
Size                      object
SizeUnitMeasureCode       object
WeightUnitMeasureCode     object
Weight                   float64
DaysToManufacture          int64
ProductLine               object
Class                     object
Style                     object
ProductSubcategoryID     float64
ProductModelID           float64
SellStartDate             object
SellEndDate               object
DiscontinuedDate         float64
rowguid                   object
ModifiedDate              object
dtype: object

**Class Question:** Why do datatypes matter? What operations could we perform on some datatypes that we could not on others? Note the importance of this in checking dataset integrity.

## Selecting a Column

We can select columns in two ways. Either we treat the column as an attribute of the DataFrame or we index the DataFrame for a specific element (in this case, the element is a column name).

In [15]:
# show the difference in data types by using single and double square brackets
prod["Name"]
# type(prod["Name"])  # series

ProductID
1                Adjustable Race
2                   Bearing Ball
3                BB Ball Bearing
4          Headset Ball Bearings
316                        Blade
317                  LL Crankarm
318                  ML Crankarm
319                  HL Crankarm
320              Chainring Bolts
321                Chainring Nut
322                    Chainring
323                   Crown Race
324                  Chain Stays
325                      Decal 1
326                      Decal 2
327                    Down Tube
328            Mountain End Caps
329                Road End Caps
330             Touring End Caps
331                     Fork End
332                    Freewheel
341                Flat Washer 1
342                Flat Washer 6
343                Flat Washer 2
344                Flat Washer 9
345                Flat Washer 4
346                Flat Washer 3
347                Flat Washer 8
348                Flat Washer 5
349                Flat Washer 7


In [16]:
prod[["Name"]]
# type(prod[["Name"]]) # dataframe

Name
ProductID                           
1                    Adjustable Race
2                       Bearing Ball
3                    BB Ball Bearing
4              Headset Ball Bearings
316                            Blade
317                      LL Crankarm
318                      ML Crankarm
319                      HL Crankarm
320                  Chainring Bolts
321                    Chainring Nut
322                        Chainring
323                       Crown Race
324                      Chain Stays
325                          Decal 1
326                          Decal 2
327                        Down Tube
328                Mountain End Caps
329                    Road End Caps
330                 Touring End Caps
331                         Fork End
332                        Freewheel
341                    Flat Washer 1
342                    Flat Washer 6
343                    Flat Washer 2
344                    Flat Washer 9
345                    Flat Washer 4
346                    Flat Washer 3
347                    Flat Washer 8
348                    Flat Washer 5
349                    Flat Washer 7
...                              ...
970            Touring-2000 Blue, 46
971            Touring-2000 Blue, 50
972            Touring-2000 Blue, 54
973            Road-350-W Yellow, 40
974            Road-350-W Yellow, 42
975            Road-350-W Yellow, 44
976            Road-350-W Yellow, 48
977               Road-750 Black, 58
978            Touring-3000 Blue, 44
979            Touring-3000 Blue, 50
980        Mountain-400-W Silver, 38
981        Mountain-400-W Silver, 40
982        Mountain-400-W Silver, 42
983        Mountain-400-W Silver, 46
984          Mountain-500 Silver, 40
985          Mountain-500 Silver, 42
986          Mountain-500 Silver, 44
987          Mountain-500 Silver, 48
988          Mountain-500 Silver, 52
989           Mountain-500 Black, 40
990           Mountain-500 Black, 42
991           Mountain-500 Black, 44
992           Mountain-500 Black, 48
993           Mountain-500 Black, 52
994                LL Bottom Bracket
995                ML Bottom Bracket
996                HL Bottom Bracket
997               Road-750 Black, 44
998               Road-750 Black, 48
999               Road-750 Black, 52

[504 rows x 1 columns]

In [18]:
# select the Name column only, returned as a Series object

In [19]:
# select the Name column only, returned as a DataFrame object

In [17]:
# selecting > 1 column (must use double brackets!)
# prod[["Name", "Color", "MakeFlag"]]
cols_to_subset = ["Name", "Color", "MakeFlag"]
prod[cols_to_subset]

Name   Color  MakeFlag
ProductID                                             
1                    Adjustable Race     NaN         0
2                       Bearing Ball     NaN         0
3                    BB Ball Bearing     NaN         1
4              Headset Ball Bearings     NaN         0
316                            Blade     NaN         1
317                      LL Crankarm   Black         0
318                      ML Crankarm   Black         0
319                      HL Crankarm   Black         0
320                  Chainring Bolts  Silver         0
321                    Chainring Nut  Silver         0
322                        Chainring   Black         0
323                       Crown Race     NaN         0
324                      Chain Stays     NaN         1
325                          Decal 1     NaN         0
326                          Decal 2     NaN         0
327                        Down Tube     NaN         1
328                Mountain End Caps     NaN         1
329                    Road End Caps     NaN         1
330                 Touring End Caps     NaN         1
331                         Fork End     NaN         1
332                        Freewheel  Silver         0
341                    Flat Washer 1     NaN         0
342                    Flat Washer 6     NaN         0
343                    Flat Washer 2     NaN         0
344                    Flat Washer 9     NaN         0
345                    Flat Washer 4     NaN         0
346                    Flat Washer 3     NaN         0
347                    Flat Washer 8     NaN         0
348                    Flat Washer 5     NaN         0
349                    Flat Washer 7     NaN         0
...                              ...     ...       ...
970            Touring-2000 Blue, 46    Blue         1
971            Touring-2000 Blue, 50    Blue         1
972            Touring-2000 Blue, 54    Blue         1
973            Road-350-W Yellow, 40  Yellow         1
974            Road-350-W Yellow, 42  Yellow         1
975            Road-350-W Yellow, 44  Yellow         1
976            Road-350-W Yellow, 48  Yellow         1
977               Road-750 Black, 58   Black         1
978            Touring-3000 Blue, 44    Blue         1
979            Touring-3000 Blue, 50    Blue         1
980        Mountain-400-W Silver, 38  Silver         1
981        Mountain-400-W Silver, 40  Silver         1
982        Mountain-400-W Silver, 42  Silver         1
983        Mountain-400-W Silver, 46  Silver         1
984          Mountain-500 Silver, 40  Silver         1
985          Mountain-500 Silver, 42  Silver         1
986          Mountain-500 Silver, 44  Silver         1
987          Mountain-500 Silver, 48  Silver         1
988          Mountain-500 Silver, 52  Silver         1
989           Mountain-500 Black, 40   Black         1
990           Mountain-500 Black, 42   Black         1
991           Mountain-500 Black, 44   Black         1
992           Mountain-500 Black, 48   Black         1
993           Mountain-500 Black, 52   Black         1
994                LL Bottom Bracket     NaN         1
995                ML Bottom Bracket     NaN         1
996                HL Bottom Bracket     NaN         1
997               Road-750 Black, 44   Black         1
998               Road-750 Black, 48   Black         1
999               Road-750 Black, 52   Black         1

[504 rows x 3 columns]

**Class Question:** What if we wanted to select a column that has a space in it? Which method from the above two would we use? Why?

In [21]:
prod.head()

Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   

          Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  \
ProductID                                                                       
1           NaN              1000           750           0.0        0.0  NaN   
2           NaN              1000           750           0.0        0.0  NaN   
3           NaN               800           600           0.0        0.0  NaN   
4           NaN               800           600           0.0        0.0  NaN   
316         NaN               800           600           0.0        0.0  NaN   

           ... ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID  ...                                                                 
1          ...         NaN   NaN    NaN                   NaN            NaN   
2          ...         NaN   NaN    NaN                   NaN            NaN   
3          ...         NaN   NaN    NaN                   NaN            NaN   
4          ...         NaN   NaN    NaN                   NaN            NaN   
316        ...         NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

In [18]:
# rename one or more columns with a dictionary. Note: inplace=False will return a new dataframe,
# but leave the original dataframe untouched. Change this to True to modify the original dataframe.
prod.rename(columns={"Name": "ProductName", "ProductNumber": "Number"}, inplace=True)

In [19]:
prod.head()

ProductName   Number  MakeFlag  FinishedGoodsFlag Color  \
ProductID                                                                      
1                Adjustable Race  AR-5381         0                  0   NaN   
2                   Bearing Ball  BA-8327         0                  0   NaN   
3                BB Ball Bearing  BE-2349         1                  0   NaN   
4          Headset Ball Bearings  BE-2908         0                  0   NaN   
316                        Blade  BL-2036         1                  0   NaN   

           SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  ...  \
ProductID                                                                ...   
1                      1000           750           0.0        0.0  NaN  ...   
2                      1000           750           0.0        0.0  NaN  ...   
3                       800           600           0.0        0.0  NaN  ...   
4                       800           600           0.0        0.0  NaN  ...   
316                     800           600           0.0        0.0  NaN  ...   

          ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID                                                                 
1                 NaN   NaN    NaN                   NaN            NaN   
2                 NaN   NaN    NaN                   NaN            NaN   
3                 NaN   NaN    NaN                   NaN            NaN   
4                 NaN   NaN    NaN                   NaN            NaN   
316               NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

Renaming **ALL** columns with a new list of column names.

Note that the `pd.DataFrame.columns` property can be cast to a `list` type. Originally, it's a `pd.core.indexes.base.Index` object:

In [24]:
# print('My columns look like:\n', prod.columns)
# print('\nAnd the type is:\n', type(prod.columns))

We can explicitly cast these to a list object as such, by using the built-in `list()` function:

In [25]:
# print('Now my columns look like:\n', list(prod.columns))
# print('\nAnd are of type:\n', type(list(prod.columns)))

We can place these columns into a variable, `cols`:

In [21]:
# declare a list of strings - these strings will become the new column names
cols = list(prod.columns)

In [22]:
cols[0] = "Name"
cols

['Name',
 'Number',
 'MakeFlag',
 'FinishedGoodsFlag',
 'Color',
 'SafetyStockLevel',
 'ReorderPoint',
 'StandardCost',
 'ListPrice',
 'Size',
 'SizeUnitMeasureCode',
 'WeightUnitMeasureCode',
 'Weight',
 'DaysToManufacture',
 'ProductLine',
 'Class',
 'Style',
 'ProductSubcategoryID',
 'ProductModelID',
 'SellStartDate',
 'SellEndDate',
 'DiscontinuedDate',
 'rowguid',
 'ModifiedDate']

In [25]:
prod.columns = cols



In [26]:
prod.head()

Name   Number  MakeFlag  FinishedGoodsFlag Color  \
ProductID                                                                      
1                Adjustable Race  AR-5381         0                  0   NaN   
2                   Bearing Ball  BA-8327         0                  0   NaN   
3                BB Ball Bearing  BE-2349         1                  0   NaN   
4          Headset Ball Bearings  BE-2908         0                  0   NaN   
316                        Blade  BL-2036         1                  0   NaN   

           SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  ...  \
ProductID                                                                ...   
1                      1000           750           0.0        0.0  NaN  ...   
2                      1000           750           0.0        0.0  NaN  ...   
3                       800           600           0.0        0.0  NaN  ...   
4                       800           600           0.0        0.0  NaN  ...   
316                     800           600           0.0        0.0  NaN  ...   

          ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID                                                                 
1                 NaN   NaN    NaN                   NaN            NaN   
2                 NaN   NaN    NaN                   NaN            NaN   
3                 NaN   NaN    NaN                   NaN            NaN   
4                 NaN   NaN    NaN                   NaN            NaN   
316               NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

We can use list indexing to mutate the columns we want:

In [31]:
# overwrite one of the column names using list indexing

Finally, we can set the `pd.DataFrame.columns` property (this is a settable class property), to the new `cols` list, overwriting the existing columns header names:

In [32]:
# Note that our first column name has changed from 'Name' to 'NewName'

In [33]:
# display the final result of the 'NewName' column
# prod.head(3)[['NewName']]

## Common Column Operations

While this is non-comprehensive, these are a few key column-specific data checks.


**Descriptive statistics:**  the minimum, first quartile, median, third quartile, and maximum.

(And more! The mean too.)

Five Number Summary (all assumes numeric data):
- **Min:** The smallest value in the column
- **Max:** The largest value in the column
- **Quartile:** A quartile is one fourth of our data
    - **First quartile:** This is the bottom most 25 percent
    - **Median:** The middle value. (Line all values biggest to smallest - median is the middle!) Also the 50th percentile
    - **Third quartile:** This the the top 75 percentile of our data


![](https://www.mathsisfun.com/data/images/quartiles-a.svg)

In [28]:
# note - .describe() *default* only checks numeric datatypes
# show .describe() for the 'MakeFlag', 'SafetyStockLevel', 'StandardCost' fields.
prod[["MakeFlag", "SafetyStockLevel", "StandardCost"]].describe()

MakeFlag  SafetyStockLevel  StandardCost
count  504.000000        504.000000    504.000000
mean     0.474206        535.150794    258.602961
std      0.499830        374.112954    461.632808
min      0.000000          4.000000      0.000000
25%      0.000000        100.000000      0.000000
50%      0.000000        500.000000     23.372200
75%      1.000000       1000.000000    317.075825
max      1.000000       1000.000000   2171.294200

**Value Counts:** `pd.Series.value_counts()` count the occurrence of each value within our series.

In [35]:
# show the most popular product colors (aggregated by count, descending by default)
prod.head()

Name1  Number1  MakeFlag1  FinishedGoodsFlag1  \
ProductID                                                                  
1                Adjustable Race  AR-5381          0                   0   
2                   Bearing Ball  BA-8327          0                   0   
3                BB Ball Bearing  BE-2349          1                   0   
4          Headset Ball Bearings  BE-2908          0                   0   
316                        Blade  BL-2036          1                   0   

          Color1  SafetyStockLevel1  ReorderPoint1  StandardCost1  ListPrice1  \
ProductID                                                                       
1            NaN               1000            750            0.0         0.0   
2            NaN               1000            750            0.0         0.0   
3            NaN                800            600            0.0         0.0   
4            NaN                800            600            0.0         0.0   
316          NaN                800            600            0.0         0.0   

          Size1  ... ProductLine1 Class1  Style1  ProductSubcategoryID1  \
ProductID        ...                                                      
1           NaN  ...          NaN    NaN     NaN                    NaN   
2           NaN  ...          NaN    NaN     NaN                    NaN   
3           NaN  ...          NaN    NaN     NaN                    NaN   
4           NaN  ...          NaN    NaN     NaN                    NaN   
316         NaN  ...          NaN    NaN     NaN                    NaN   

          ProductModelID1       SellStartDate1 SellEndDate1  \
ProductID                                                     
1                     NaN  2008-04-30 00:00:00          NaN   
2                     NaN  2008-04-30 00:00:00          NaN   
3                     NaN  2008-04-30 00:00:00          NaN   
4                     NaN  2008-04-30 00:00:00          NaN   
316                   NaN  2008-04-30 00:00:00          NaN   

           DiscontinuedDate1                                rowguid1  \
ProductID                                                              
1                        NaN  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2                        NaN  {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3                        NaN  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4                        NaN  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316                      NaN  {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                           ModifiedDate1  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

In [29]:
prod["Color"].value_counts().sum()

256

In [30]:
prod["Color"].unique()

array([nan, 'Black', 'Silver', 'Red', 'White', 'Blue', 'Multi', 'Yellow',
       'Grey', 'Silver/Black'], dtype=object)

**Unique values:** Determine the number of distinct values within a given series.

In [30]:
# What are the unique colors for the products?

In [31]:
# how many distinct colors are there?

In [32]:
# We can also include nulls with .nunique() as such:

## Filtering on a Single Condition

Filtering and sorting are key processes that allow us to drill into the 'nitty gritty' and cross sections of our dataset.

To filter, we use a process called **Boolean Filtering**, wherein we define a Boolean condition, and use that Boolean condition to filer on our DataFrame.

Recall: our given dataset has a column `Color`. Let's see if we can find all products that are `Black`. Let's take a look at the first 10 rows of the dataframe to see how it looks as-is:

In [31]:
# show the first 10 rows of the 'Color' column
prod.head(10)

Name   Number  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                
1                Adjustable Race  AR-5381         0                  0   
2                   Bearing Ball  BA-8327         0                  0   
3                BB Ball Bearing  BE-2349         1                  0   
4          Headset Ball Bearings  BE-2908         0                  0   
316                        Blade  BL-2036         1                  0   
317                  LL Crankarm  CA-5965         0                  0   
318                  ML Crankarm  CA-6738         0                  0   
319                  HL Crankarm  CA-7457         0                  0   
320              Chainring Bolts  CB-2903         0                  0   
321                Chainring Nut  CN-6137         0                  0   

            Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                    
1             NaN              1000           750           0.0        0.0   
2             NaN              1000           750           0.0        0.0   
3             NaN               800           600           0.0        0.0   
4             NaN               800           600           0.0        0.0   
316           NaN               800           600           0.0        0.0   
317         Black               500           375           0.0        0.0   
318         Black               500           375           0.0        0.0   
319         Black               500           375           0.0        0.0   
320        Silver              1000           750           0.0        0.0   
321        Silver              1000           750           0.0        0.0   

          Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
1          NaN  ...         NaN   NaN    NaN                   NaN   
2          NaN  ...         NaN   NaN    NaN                   NaN   
3          NaN  ...         NaN   NaN    NaN                   NaN   
4          NaN  ...         NaN   NaN    NaN                   NaN   
316        NaN  ...         NaN   NaN    NaN                   NaN   
317        NaN  ...         NaN    L     NaN                   NaN   
318        NaN  ...         NaN    M     NaN                   NaN   
319        NaN  ...         NaN   NaN    NaN                   NaN   
320        NaN  ...         NaN   NaN    NaN                   NaN   
321        NaN  ...         NaN   NaN    NaN                   NaN   

          ProductModelID        SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                                     
1                    NaN  2008-04-30 00:00:00         NaN               NaN   
2                    NaN  2008-04-30 00:00:00         NaN               NaN   
3                    NaN  2008-04-30 00:00:00         NaN               NaN   
4                    NaN  2008-04-30 00:00:00         NaN               NaN   
316                  NaN  2008-04-30 00:00:00         NaN               NaN   
317                  NaN  2008-04-30 00:00:00         NaN               NaN   
318                  NaN  2008-04-30 00:00:00         NaN               NaN   
319                  NaN  2008-04-30 00:00:00         NaN               NaN   
320                  NaN  2008-04-30 00:00:00         NaN               NaN   
321                  NaN  2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   
317        {3C9D10B7-A6B2-4774-9963-C19DC

In [32]:
prod[["Color"]].head(10)

Color
ProductID        
1             NaN
2             NaN
3             NaN
4             NaN
316           NaN
317         Black
318         Black
319         Black
320        Silver
321        Silver

By applying a `boolean mask` to this dataframe, `== 'Black'`, we can get the following:

In [34]:
# set the previous output to == Black
prod[prod["Color"] == "Black"]

Name      Number  MakeFlag  \
ProductID                                                        
317                          LL Crankarm     CA-5965         0   
318                          ML Crankarm     CA-6738         0   
319                          HL Crankarm     CA-7457         0   
322                            Chainring     CR-7833         0   
680            HL Road Frame - Black, 58  FR-R92B-58         1   
708              Sport-100 Helmet, Black     HL-U509         0   
722            LL Road Frame - Black, 58  FR-R38B-58         1   
723            LL Road Frame - Black, 60  FR-R38B-60         1   
724            LL Road Frame - Black, 62  FR-R38B-62         1   
736            LL Road Frame - Black, 44  FR-R38B-44         1   
737            LL Road Frame - Black, 48  FR-R38B-48         1   
738            LL Road Frame - Black, 52  FR-R38B-52         1   
743        HL Mountain Frame - Black, 42  FR-M94B-42         1   
744        HL Mountain Frame - Black, 44  FR-M94B-44         1   
745        HL Mountain Frame - Black, 48  FR-M94B-48         1   
746        HL Mountain Frame - Black, 46  FR-M94B-46         1   
747        HL Mountain Frame - Black, 38  FR-M94B-38         1   
765                   Road-650 Black, 58  BK-R50B-58         1   
766                   Road-650 Black, 60  BK-R50B-60         1   
767                   Road-650 Black, 62  BK-R50B-62         1   
768                   Road-650 Black, 44  BK-R50B-44         1   
769                   Road-650 Black, 48  BK-R50B-48         1   
770                   Road-650 Black, 52  BK-R50B-52         1   
775               Mountain-100 Black, 38  BK-M82B-38         1   
776               Mountain-100 Black, 42  BK-M82B-42         1   
777               Mountain-100 Black, 44  BK-M82B-44         1   
778               Mountain-100 Black, 48  BK-M82B-48         1   
782               Mountain-200 Black, 38  BK-M68B-38         1   
783               Mountain-200 Black, 42  BK-M68B-42         1   
784               Mountain-200 Black, 46  BK-M68B-46         1   
...                                  ...         ...       ...   
851              Men's Sports Shorts, XL   SH-M897-X         0   
852                    Women's Tights, S   TG-W091-S         0   
853                    Women's Tights, M   TG-W091-M         0   
854                    Women's Tights, L   TG-W091-L         0   
858                Half-Finger Gloves, S   GL-H102-S         0   
859                Half-Finger Gloves, M   GL-H102-M         0   
860                Half-Finger Gloves, L   GL-H102-L         0   
861                Full-Finger Gloves, S   GL-F110-S         0   
862                Full-Finger Gloves, M   GL-F110-M         0   
863                Full-Finger Gloves, L   GL-F110-L         0   
867           Women's Mountain Shorts, S   SH-W890-S         0   
868           Women's Mountain Shorts, M   SH-W890-M         0   
869           Women's Mountain Shorts, L   SH-W890-L         0   
924        LL Mountain Frame - Black, 42  FR-M21B-42         1   
925        LL Mountain Frame - Black, 44  FR-M21B-44         1   
926        LL Mountain Frame - Black, 48  FR-M21B-48         1   
927        LL Mountain Frame - Black, 52  FR-M21B-52         1   
943        LL Mountain Frame - Black, 40  FR-M21B-40         1   
949                          LL Crankset     CS-4759         1   
950                          ML Crankset     CS-6583         1   
951                          HL Crankset     CS-9183         1   
977                   Road-750 Black, 58  BK-R19B-58         1   
989               Mountain-500 Black, 40  BK-M18B-40         1   
990               Mountain-500 Black, 42  BK-M18B-42         1   
991               Mountain-500 Black, 44  BK-M18B-44         1   
992               Mountain-500 Black, 48  BK-M18B-48         1   
993               Mountain-500 Black, 52  BK-M18B-52         1   
997                   Road-750 Black, 44  BK-R19B-44         1   
998       

Now we can use that 'mask' from above, and apply it to our full dataframe. Every time we have a `True` in a row, we return the row. If we have a `False` in that row, we do not return it. The result is a dataframe that only has rows where `Color` is `Black`:

In [36]:
# show the full prod DataFrame where Color is Black
prod[prod["Color"] == "Black"][["Name", "Number"]]

Name      Number
ProductID                                           
317                          LL Crankarm     CA-5965
318                          ML Crankarm     CA-6738
319                          HL Crankarm     CA-7457
322                            Chainring     CR-7833
680            HL Road Frame - Black, 58  FR-R92B-58
708              Sport-100 Helmet, Black     HL-U509
722            LL Road Frame - Black, 58  FR-R38B-58
723            LL Road Frame - Black, 60  FR-R38B-60
724            LL Road Frame - Black, 62  FR-R38B-62
736            LL Road Frame - Black, 44  FR-R38B-44
737            LL Road Frame - Black, 48  FR-R38B-48
738            LL Road Frame - Black, 52  FR-R38B-52
743        HL Mountain Frame - Black, 42  FR-M94B-42
744        HL Mountain Frame - Black, 44  FR-M94B-44
745        HL Mountain Frame - Black, 48  FR-M94B-48
746        HL Mountain Frame - Black, 46  FR-M94B-46
747        HL Mountain Frame - Black, 38  FR-M94B-38
765                   Road-650 Black, 58  BK-R50B-58
766                   Road-650 Black, 60  BK-R50B-60
767                   Road-650 Black, 62  BK-R50B-62
768                   Road-650 Black, 44  BK-R50B-44
769                   Road-650 Black, 48  BK-R50B-48
770                   Road-650 Black, 52  BK-R50B-52
775               Mountain-100 Black, 38  BK-M82B-38
776               Mountain-100 Black, 42  BK-M82B-42
777               Mountain-100 Black, 44  BK-M82B-44
778               Mountain-100 Black, 48  BK-M82B-48
782               Mountain-200 Black, 38  BK-M68B-38
783               Mountain-200 Black, 42  BK-M68B-42
784               Mountain-200 Black, 46  BK-M68B-46
...                                  ...         ...
851              Men's Sports Shorts, XL   SH-M897-X
852                    Women's Tights, S   TG-W091-S
853                    Women's Tights, M   TG-W091-M
854                    Women's Tights, L   TG-W091-L
858                Half-Finger Gloves, S   GL-H102-S
859                Half-Finger Gloves, M   GL-H102-M
860                Half-Finger Gloves, L   GL-H102-L
861                Full-Finger Gloves, S   GL-F110-S
862                Full-Finger Gloves, M   GL-F110-M
863                Full-Finger Gloves, L   GL-F110-L
867           Women's Mountain Shorts, S   SH-W890-S
868           Women's Mountain Shorts, M   SH-W890-M
869           Women's Mountain Shorts, L   SH-W890-L
924        LL Mountain Frame - Black, 42  FR-M21B-42
925        LL Mountain Frame - Black, 44  FR-M21B-44
926        LL Mountain Frame - Black, 48  FR-M21B-48
927        LL Mountain Frame - Black, 52  FR-M21B-52
943        LL Mountain Frame - Black, 40  FR-M21B-40
949                          LL Crankset     CS-4759
950                          ML Crankset     CS-6583
951                          HL Crankset     CS-9183
977                   Road-750 Black, 58  BK-R19B-58
989               Mountain-500 Black, 40  BK-M18B-40
990               Mountain-500 Black, 42  BK-M18B-42
991               Mountain-500 Black, 44  BK-M18B-44
992               Mountain-500 Black, 48  BK-M18B-48
993               Mountain-500 Black, 52  BK-M18B-52
997                   Road-750 Black, 44  BK-R19B-44
998                   Road-750 Black, 48  BK-R19B-48
999                   Road-750 Black, 52  BK-R19B-52

[93 rows x 2 columns]

Let's calculate the **average ListPrice** for the **salable products**.

> Think: What are the component parts of this problem?

In [37]:
# First, we need to get salable items. 
# Use your data dictionary from the beginning of this lesson.
salable = prod[prod["FinishedGoodsFlag"] == 1]

In [48]:
prod.head()

Name1  Number1  MakeFlag1  FinishedGoodsFlag1  \
ProductID                                                                  
1                Adjustable Race  AR-5381          0                   0   
2                   Bearing Ball  BA-8327          0                   0   
3                BB Ball Bearing  BE-2349          1                   0   
4          Headset Ball Bearings  BE-2908          0                   0   
316                        Blade  BL-2036          1                   0   

          Color1  SafetyStockLevel1  ReorderPoint1  StandardCost1  ListPrice1  \
ProductID                                                                       
1            NaN               1000            750            0.0         0.0   
2            NaN               1000            750            0.0         0.0   
3            NaN                800            600            0.0         0.0   
4            NaN                800            600            0.0         0.0   
316          NaN                800            600            0.0         0.0   

          Size1  ... ProductLine1 Class1  Style1  ProductSubcategoryID1  \
ProductID        ...                                                      
1           NaN  ...          NaN    NaN     NaN                    NaN   
2           NaN  ...          NaN    NaN     NaN                    NaN   
3           NaN  ...          NaN    NaN     NaN                    NaN   
4           NaN  ...          NaN    NaN     NaN                    NaN   
316         NaN  ...          NaN    NaN     NaN                    NaN   

          ProductModelID1       SellStartDate1 SellEndDate1  \
ProductID                                                     
1                     NaN  2008-04-30 00:00:00          NaN   
2                     NaN  2008-04-30 00:00:00          NaN   
3                     NaN  2008-04-30 00:00:00          NaN   
4                     NaN  2008-04-30 00:00:00          NaN   
316                   NaN  2008-04-30 00:00:00          NaN   

           DiscontinuedDate1                                rowguid1  \
ProductID                                                              
1                        NaN  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2                        NaN  {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3                        NaN  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4                        NaN  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316                      NaN  {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                           ModifiedDate1  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

Now, we need to find average list price of those above items. Let's just get the 'ListPrice' column for starters.

In [39]:
salable["ListPrice"].describe()

count     295.000000
mean      744.595220
std       892.563172
min         2.290000
25%        66.745000
50%       337.220000
75%      1100.240000
max      3578.270000
Name: ListPrice, dtype: float64

To get the average of that column, just take `.mean()`

In [38]:
salable[salable["ListPrice"] > 3000]

Name      Number  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
749               Road-150 Red, 62  BK-R93R-62         1                  1   
750               Road-150 Red, 44  BK-R93R-44         1                  1   
751               Road-150 Red, 48  BK-R93R-48         1                  1   
752               Road-150 Red, 52  BK-R93R-52         1                  1   
753               Road-150 Red, 56  BK-R93R-56         1                  1   
771        Mountain-100 Silver, 38  BK-M82S-38         1                  1   
772        Mountain-100 Silver, 42  BK-M82S-42         1                  1   
773        Mountain-100 Silver, 44  BK-M82S-44         1                  1   
774        Mountain-100 Silver, 48  BK-M82S-48         1                  1   
775         Mountain-100 Black, 38  BK-M82B-38         1                  1   
776         Mountain-100 Black, 42  BK-M82B-42         1                  1   
777         Mountain-100 Black, 44  BK-M82B-44         1                  1   
778         Mountain-100 Black, 48  BK-M82B-48         1                  1   

            Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  \
ProductID                                                                    
749           Red               100            75     2171.2942    3578.27   
750           Red               100            75     2171.2942    3578.27   
751           Red               100            75     2171.2942    3578.27   
752           Red               100            75     2171.2942    3578.27   
753           Red               100            75     2171.2942    3578.27   
771        Silver               100            75     1912.1544    3399.99   
772        Silver               100            75     1912.1544    3399.99   
773        Silver               100            75     1912.1544    3399.99   
774        Silver               100            75     1912.1544    3399.99   
775         Black               100            75     1898.0944    3374.99   
776         Black               100            75     1898.0944    3374.99   
777         Black               100            75     1898.0944    3374.99   
778         Black               100            75     1898.0944    3374.99   

          Size  ... ProductLine Class  Style  ProductSubcategoryID  \
ProductID       ...                                                  
749         62  ...          R     H      U                    2.0   
750         44  ...          R     H      U                    2.0   
751         48  ...          R     H      U                    2.0   
752         52  ...          R     H      U                    2.0   
753         56  ...          R     H      U                    2.0   
771         38  ...          M     H      U                    1.0   
772         42  ...          M     H      U                    1.0   
773         44  ...          M     H      U                    1.0   
774         48  ...          M     H      U                    1.0   
775         38  ...          M     H      U                    1.0   
776         42  ...          M     H      U                    1.0   
777         44  ...          M     H      U                    1.0   
778         48  ...          M     H      U                    1.0   

          ProductModelID        SellStartDate          SellEndDate  \
ProductID                                                            
749                 25.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
750                 25.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
751                 25.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
752                 25.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
753                 25.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
771                 19.0  2011-05-31 00:00:00  2012-05-29 00:00:00   
772                 19.0  2011-05-31 00:00:00  2012-05-29 00:0

We can take a shortcut and just use `.describe()` here:

**Sneak peek**: Another handy trick is to use `.hist()` to get a distribution of a continuous variable - in this case, `ListPrice`. We'll cover this more in future lessons:

In [40]:
salable["ListPrice"].hist()

## Filtering on Multiple Conditions

Here, we will filter on _multiple conditions_. Before, we filtered on rows where Color was Black. We also filtered where FinishedGoodsFlag was equal to 1. Let's see what happens when we filter on *both* simultaneously. 

The format for multiple conditions is:

`df[ (df['col1'] == value1) & (df['col2'] == value2) ]`

Or, more simply:

`df[ (CONDITION 1) & (CONDITION 2) ]`

Which eventually may evaluate to something like:

`df[ True & False ]`

...on a row-by-row basis. If the end result is `False`, the row is omitted.

_Don't forget parentheses in your conditions!_ This is a common mistake.

In [41]:
# Let's look at a table where Color is Black, AND FinishedGoodsFlag is 1
black = prod["Color"] == "Black"
finished = prod["FinishedGoodsFlag"] == 1
prod[(black) & (finished)]

Name      Number  MakeFlag  \
ProductID                                                        
680            HL Road Frame - Black, 58  FR-R92B-58         1   
708              Sport-100 Helmet, Black     HL-U509         0   
722            LL Road Frame - Black, 58  FR-R38B-58         1   
723            LL Road Frame - Black, 60  FR-R38B-60         1   
724            LL Road Frame - Black, 62  FR-R38B-62         1   
736            LL Road Frame - Black, 44  FR-R38B-44         1   
737            LL Road Frame - Black, 48  FR-R38B-48         1   
738            LL Road Frame - Black, 52  FR-R38B-52         1   
743        HL Mountain Frame - Black, 42  FR-M94B-42         1   
744        HL Mountain Frame - Black, 44  FR-M94B-44         1   
745        HL Mountain Frame - Black, 48  FR-M94B-48         1   
746        HL Mountain Frame - Black, 46  FR-M94B-46         1   
747        HL Mountain Frame - Black, 38  FR-M94B-38         1   
765                   Road-650 Black, 58  BK-R50B-58         1   
766                   Road-650 Black, 60  BK-R50B-60         1   
767                   Road-650 Black, 62  BK-R50B-62         1   
768                   Road-650 Black, 44  BK-R50B-44         1   
769                   Road-650 Black, 48  BK-R50B-48         1   
770                   Road-650 Black, 52  BK-R50B-52         1   
775               Mountain-100 Black, 38  BK-M82B-38         1   
776               Mountain-100 Black, 42  BK-M82B-42         1   
777               Mountain-100 Black, 44  BK-M82B-44         1   
778               Mountain-100 Black, 48  BK-M82B-48         1   
782               Mountain-200 Black, 38  BK-M68B-38         1   
783               Mountain-200 Black, 42  BK-M68B-42         1   
784               Mountain-200 Black, 46  BK-M68B-46         1   
785               Mountain-300 Black, 38  BK-M47B-38         1   
786               Mountain-300 Black, 40  BK-M47B-40         1   
787               Mountain-300 Black, 44  BK-M47B-44         1   
788               Mountain-300 Black, 48  BK-M47B-48         1   
...                                  ...         ...       ...   
851              Men's Sports Shorts, XL   SH-M897-X         0   
852                    Women's Tights, S   TG-W091-S         0   
853                    Women's Tights, M   TG-W091-M         0   
854                    Women's Tights, L   TG-W091-L         0   
858                Half-Finger Gloves, S   GL-H102-S         0   
859                Half-Finger Gloves, M   GL-H102-M         0   
860                Half-Finger Gloves, L   GL-H102-L         0   
861                Full-Finger Gloves, S   GL-F110-S         0   
862                Full-Finger Gloves, M   GL-F110-M         0   
863                Full-Finger Gloves, L   GL-F110-L         0   
867           Women's Mountain Shorts, S   SH-W890-S         0   
868           Women's Mountain Shorts, M   SH-W890-M         0   
869           Women's Mountain Shorts, L   SH-W890-L         0   
924        LL Mountain Frame - Black, 42  FR-M21B-42         1   
925        LL Mountain Frame - Black, 44  FR-M21B-44         1   
926        LL Mountain Frame - Black, 48  FR-M21B-48         1   
927        LL Mountain Frame - Black, 52  FR-M21B-52         1   
943        LL Mountain Frame - Black, 40  FR-M21B-40         1   
949                          LL Crankset     CS-4759         1   
950                          ML Crankset     CS-6583         1   
951                          HL Crankset     CS-9183         1   
977                   Road-750 Black, 58  BK-R19B-58         1   
989               Mountain-500 Black, 40  BK-M18B-40         1   
990               Mountain-500 Black, 42  BK-M18B-42         1   
991               Mountain-500 Black, 44  BK-M18B-44         1   
992               Mountain-500 Black, 48  BK-M18B-48         1   
993               Mountain-500 Black, 52  BK-M18B-52         1   
997                   Road-750 Black, 44  BK-R19B-44         1   
998       

In [42]:
# Here we have an example of a list price of greater than 50, 
# OR a product size that is not equal to 'XL'
greater_than_50 = prod["ListPrice"] > 50
not_XL = prod["Size"] != "XL"

prod[(greater_than_50) | (not_XL)]

Name      Number  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                       
1                    Adjustable Race     AR-5381         0                  0   
2                       Bearing Ball     BA-8327         0                  0   
3                    BB Ball Bearing     BE-2349         1                  0   
4              Headset Ball Bearings     BE-2908         0                  0   
316                            Blade     BL-2036         1                  0   
317                      LL Crankarm     CA-5965         0                  0   
318                      ML Crankarm     CA-6738         0                  0   
319                      HL Crankarm     CA-7457         0                  0   
320                  Chainring Bolts     CB-2903         0                  0   
321                    Chainring Nut     CN-6137         0                  0   
322                        Chainring     CR-7833         0                  0   
323                       Crown Race     CR-9981         0                  0   
324                      Chain Stays     CS-2812         1                  0   
325                          Decal 1     DC-8732         0                  0   
326                          Decal 2     DC-9824         0                  0   
327                        Down Tube     DT-2377         1                  0   
328                Mountain End Caps     EC-M092         1                  0   
329                    Road End Caps     EC-R098         1                  0   
330                 Touring End Caps     EC-T209         1                  0   
331                         Fork End     FE-3760         1                  0   
332                        Freewheel     FH-2981         0                  0   
341                    Flat Washer 1     FW-1000         0                  0   
342                    Flat Washer 6     FW-1200         0                  0   
343                    Flat Washer 2     FW-1400         0                  0   
344                    Flat Washer 9     FW-3400         0                  0   
345                    Flat Washer 4     FW-3800         0                  0   
346                    Flat Washer 3     FW-5160         0                  0   
347                    Flat Washer 8     FW-5800         0                  0   
348                    Flat Washer 5     FW-7160         0                  0   
349                    Flat Washer 7     FW-9160         0                  0   
...                              ...         ...       ...                ...   
970            Touring-2000 Blue, 46  BK-T44U-46         1                  1   
971            Touring-2000 Blue, 50  BK-T44U-50         1                  1   
972            Touring-2000 Blue, 54  BK-T44U-54         1                  1   
973            Road-350-W Yellow, 40  BK-R79Y-40         1                  1   
974            Road-350-W Yellow, 42  BK-R79Y-42         1                  1   
975            Road-350-W Yellow, 44  BK-R79Y-44         1                  1   
976            Road-350-W Yellow, 48  BK-R79Y-48         1                  1   
977               Road-750 Black, 58  BK-R19B-58         1                  1   
978            Touring-3000 Blue, 44  BK-T18U-44         1                  1   
979            Touring-3000 Blue, 50  BK-T18U-50         1                  1   
980        Mountain-400-W Silver, 38  BK-M38S-38         1                  1   
981        Mountain-400-W Silver, 40  BK-M38S-40         1                  1   
982        Mountain-400-W Silver, 42  BK-M38S-42         1                  1   
983        Mountain-400-W Silver, 46  BK-M38S-46         1                  1   
984          Mountain-500 Silver, 40  BK-M18S-40         1                  1   
985          Mountain-500 Silver, 42  BK-M18S-42         1                  1   
986          Mountain-500 Silver, 44  BK-M18S-44         1     

## Sorting

We can sort one column of our DataFrame as well.

In [43]:
# let's sort by standard cost, descending
prod.sort_values(by="StandardCost", ascending=False)

Name      Number  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
749               Road-150 Red, 62  BK-R93R-62         1                  1   
750               Road-150 Red, 44  BK-R93R-44         1                  1   
751               Road-150 Red, 48  BK-R93R-48         1                  1   
752               Road-150 Red, 52  BK-R93R-52         1                  1   
753               Road-150 Red, 56  BK-R93R-56         1                  1   
774        Mountain-100 Silver, 48  BK-M82S-48         1                  1   
771        Mountain-100 Silver, 38  BK-M82S-38         1                  1   
772        Mountain-100 Silver, 42  BK-M82S-42         1                  1   
773        Mountain-100 Silver, 44  BK-M82S-44         1                  1   
775         Mountain-100 Black, 38  BK-M82B-38         1                  1   
777         Mountain-100 Black, 44  BK-M82B-44         1                  1   
776         Mountain-100 Black, 42  BK-M82B-42         1                  1   
778         Mountain-100 Black, 48  BK-M82B-48         1                  1   
792               Road-250 Red, 58  BK-R89R-58         1                  1   
793             Road-250 Black, 44  BK-R89B-44         1                  1   
794             Road-250 Black, 48  BK-R89B-48         1                  1   
796             Road-250 Black, 58  BK-R89B-58         1                  1   
795             Road-250 Black, 52  BK-R89B-52         1                  1   
790               Road-250 Red, 48  BK-R89R-48         1                  1   
789               Road-250 Red, 44  BK-R89R-44         1                  1   
791               Road-250 Red, 52  BK-R89R-52         1                  1   
966          Touring-1000 Blue, 46  BK-T79U-46         1                  1   
969          Touring-1000 Blue, 60  BK-T79U-60         1                  1   
954        Touring-1000 Yellow, 46  BK-T79Y-46         1                  1   
955        Touring-1000 Yellow, 50  BK-T79Y-50         1                  1   
956        Touring-1000 Yellow, 54  BK-T79Y-54         1                  1   
968          Touring-1000 Blue, 54  BK-T79U-54         1                  1   
957        Touring-1000 Yellow, 60  BK-T79Y-60         1                  1   
967          Touring-1000 Blue, 50  BK-T79U-50         1                  1   
781        Mountain-200 Silver, 46  BK-M68S-46         1                  1   
...                            ...         ...       ...                ...   
430            Thin-Jam Lock Nut 3     LJ-5161         0                  0   
432           Thin-Jam Lock Nut 13     LJ-5811         0                  0   
405         External Lock Washer 9     LE-1201         0                  0   
433           Thin-Jam Lock Nut 14     LJ-5818         0                  0   
434            Thin-Jam Lock Nut 7     LJ-7161         0                  0   
435            Thin-Jam Lock Nut 8     LJ-7162         0                  0   
436           Thin-Jam Lock Nut 12     LJ-9080         0                  0   
437           Thin-Jam Lock Nut 11     LJ-9161         0                  0   
438                     Lock Nut 5     LN-1024         0                  0   
439                     Lock Nut 6     LN-1032         0                  0   
424            Thin-Jam Lock Nut 1     LJ-1420         0                  0   
423           Thin-Jam Lock Nut 10     LJ-1220         0                  0   
422            Thin-Jam Lock Nut 9     LJ-1213         0                  0   
421         Internal Lock Washer 2     LI-8000         0                  0   
406         External Lock Washer 5     LE-1400         0                  0   
407         External Lock Washer 7     LE-3800         0                  0   
408         External Lock Washer 6     LE-5160         0                  0   
409         External Lock Washer 1     LE-6000         0                  0   
4

This one is a little more advanced, but it demonstrates a few things:
- Conversion of a `numpy.ndarray` object (return type of `pd.Series.unique()`) into a `pd.Series` object
- `pd.Series.sort_values` with the `by=` kwarg omitted (if only one column is the operand, `by=` doesn't need specified
- Alphabetical sort of a string field, `ascending=True` means A->Z
- Inclusion of nulls, `NaN` in a string field (versus omission with a float/int as prior example)

## Independent Exercises

Do your best to complete the following prompts. Don't hesitate to look at code we wrote together!

Print the first 4 rows of the whole DataFrame.

In [44]:
# your answer here
prod.head(4)

Name   Number  MakeFlag  FinishedGoodsFlag Color  \
ProductID                                                                      
1                Adjustable Race  AR-5381         0                  0   NaN   
2                   Bearing Ball  BA-8327         0                  0   NaN   
3                BB Ball Bearing  BE-2349         1                  0   NaN   
4          Headset Ball Bearings  BE-2908         0                  0   NaN   

           SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  ...  \
ProductID                                                                ...   
1                      1000           750           0.0        0.0  NaN  ...   
2                      1000           750           0.0        0.0  NaN  ...   
3                       800           600           0.0        0.0  NaN  ...   
4                       800           600           0.0        0.0  NaN  ...   

          ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID                                                                 
1                 NaN   NaN    NaN                   NaN            NaN   
2                 NaN   NaN    NaN                   NaN            NaN   
3                 NaN   NaN    NaN                   NaN            NaN   
4                 NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  

[4 rows x 24 columns]

How many rows are in the dataframe? Return the answer as an int.

In [45]:
# your answer here
prod.shape[0]

504

How many columns? Retrun the answer as an int.

In [46]:
# your answer here
prod.shape[1]

24

How many different product lines are there?

In [81]:
# your answer here
prod["ProductLine"].nunique()

4

What are the values of these product lines?

In [82]:
# your answer here
prod["ProductLine"].unique()

array([nan, 'R ', 'S ', 'M ', 'T '], dtype=object)

Do the number of values for the product lines match the number you have using `.nunique()`? Why or why not?

In [83]:
# your answer here
# no since nan is in the array

Take the output from your previous answer (using `.unique()`). Select the label corresponding to the `Road` product line using list indexing notation. How many characters are in this string?

In [84]:
# your answer here
prod["ProductLine"].unique()[1]
len(prod["ProductLine"].unique()[1])

2

Do you notice anything odd about this?

In [85]:
# your answer here
# there's a space after it

How many products are there for the `Road` product line? Don't forget what you just worked on above! Return your answer as an int.

In [88]:
# your answer here
prod[prod["ProductLine"] == "R "].shape

(100, 24)

How many products are there in the `Women's` `Mountain` category? Return your answer as an int. _Hint: Use the data dictionary above!_

In [96]:
# your answer here
mountain_mask = prod["ProductLine"] == "M "
women_mask = prod["Style"] == "W "
prod[mountain_mask & women_mask]
prod[mountain_mask & women_mask].shape

(11, 24)

**Challenge:** What are the top 3 _most expensive list price_ product that are either in the `Women's` `Mountain` category, _OR_ `Silver` in `Color`? Return your answer as a DataFrame object, with the `ProductID` index, `NewName` relabeled as `Name`, and `ListPrice` columns. Perform the statement in one execution, and do not mutate the source DataFrame.

In [101]:
# your answer here
silver_mask = prod["Color"] == "Silver"
prod[silver_mask].shape

(43, 24)

In [107]:
prod[(mountain_mask & women_mask) | silver_mask].sort_values(by="ListPrice", ascending=False).head(3)[["Name", "ListPrice"]]

Name  ListPrice
ProductID                                    
774        Mountain-100 Silver, 48    3399.99
771        Mountain-100 Silver, 38    3399.99
772        Mountain-100 Silver, 42    3399.99

In [108]:
prod[(mountain_mask & women_mask) | silver_mask].sort_values(by="ListPrice", ascending=False).tail(3)[["Name", "ListPrice"]]

Name  ListPrice
ProductID                                  
351        Front Derailleur Cage        0.0
332                    Freewheel        0.0
320              Chainring Bolts        0.0

## Recap

We covered a lot of ground! It's ok if this takes a while to gel.

```python

# basic DataFrame operations
df.head()
df.tail()
df.shape
df.columns
df.Index

# selecting columns
df.column_name
df['column_name']

# renaming columns
df.rename({'old_name':'new_name'}, inplace=True)
df.columns = ['new_column_a', 'new_column_b']

# notable columns operations
df.describe() # five number summary
df['col1'].nunique() # number of unique values
df['col1'].value_counts() # number of occurrences of each value in column

# filtering
df[ df['col1'] < 50 ] # filter column to be less than 50
df[ (df['col1'] == value1) & (df['col2'] > value2) ] # filter column where col1 is equal to value1 AND col2 is greater to value 2

# sorting
df.sort_values(by='column_name', ascending = False) # sort biggest to smallest

```


It's common to refer back to your own code *all the time.* Don't hesistate to reference this guide! 🐼


